### ISSUE TITLE
HELP WANTED: Lets improve the RxDB documentation

### ISSUE THREAD

**Issue by pubkey** [pubkey/rxdb#6473]
In the RxDB user survey 2024 people voted that the documentation is the biggest drawback of RxDB.

So the plan is to rework all documentation pages in the next few weeks to make the better appealing to people that learn RxDB. And for that I need your help. Please give me some hints about:

- Which parts of the documentation are confusion to you?
- What was the hardest to learn when your started using RxDB?
- Which information is missing?

Please give me some feedback. You do not have to write the docs by yourself, just some hints are enough. You do not need to make big explanations.

Thank you 🙏

**Comment by jyling** [comment]
Hi, I used to struggle a lot on the Conflict handler before based on this link
https://rxdb.info/transactions-conflicts-revisions.html#custom-conflict-handler

While the basic function gave me an idea of how it would look, I'm unable to find the documentation about the conflict handler's second argument, which is remote. 

it would be nice to have an explanation and sample on the second argument of the conflict handler which is the remote, since the conflict handler could be called many times.

so far I am aware that these are the few possible values the remote argument will provide based on the RxDB source code. but a detailed explanation of each of the values will be helpful for newcomers and existing ones.

- replication-resolve-conflict
- upstream-check-if-equal
- downstream-check-if-equal-0
- downstream-check-if-equal-1


of course, It could also be my mistake, the documentation is there but I just didn't see it. 


Thank you.

**Comment by troywweber7** [comment]
This is minor, but I think one thing I and my company struggled on a lot was our first time setting up replication. I wouldn't say the documentation was bad, it was more that some of the documentation examples felt incomplete or maybe had typos that caused a little confusion when trying to connect the dots for the first time. It also took me a while to realize that before reading the graphql replication documentation, I should probably have read the general replication documentation first.

My company still needs to dig into better conflict handling, schema versioning and migration, and paid plugins, but overall we intend to continue using RxDB.

**Comment by enrico-teterra-maramalabs** [comment]
Hey! I've been working with migrations recently and found the [migration-schema documentation](https://rxdb.info/migration-schema.html) very helpful. All the information is on the page there, but it's difficult to understand what is relevant and what to skip.

I think the structure could be improved to make the key information more accessible. For example I initially skipped over sections like autoMigrate and migrationStates because I wasn’t sure what they were or why they were important

I also think "migration strategies" sound complication and could maybe just be called migration functions, because that's what they are at the end of the day

Here’s an idea how the page might be restructured:

- Introduction
- Installation 
- Getting Started

Advanced Use Cases
- Async migrations
- Filtering documents
- Migrating attachments

Choosing When the Migration Runs
- Automatic (default)
- Manual

Frequent Questions
- Do migrations work with multiple tabs?
- Do migrations work with replication?

I made the following assumptions:
- most people will only read up to getting started
- a few people will read about any important choices they need to make
- return visitors will read a few more sections, but only if they seem relevant to a question or problem they have

**Comment by robinduerhager** [comment]
I just looked at the docs to see if RxDB would fit my needs. During my research for using `attachments` I saw that neither in the `Attachments` section nor in the `Replication Protocol` section, there was a table that states which replication plugin (or which overall configuration) supports attachments.

Since it states in `Replication Protocol` that "However not all replication plugins support it.", i think this would be beneficial to immediately identify which plugins (or configuration in general) would fit ones needs :).

**Comment by mattriese** [comment]
I would appreciate pretty diagrams like some other options have. It both makes it easier to orient your thinking when learning how the library works, it also makes the project seem more professional and legitimate. 

for example, see all the diagrams on this page: https://docs.powersync.com/overview/readme-1

**Comment by fadnincx** [comment]
I would appreciate multiple layers of depth of the documentation of the replication. As I feel this is where one often interacts with custom code/other software that needs to be adopted to rxdb. Especially since each replication plugin is different.

For me, for example having a custom webrtc replication endpoint (in golang) requires a lot of reverse engineering on all the small implementation details before coming even close to the concepts mentioned in the documentation.

**Comment by tsekiguchi** [comment]
This is a worthy project, and I commend you for taking this on! I love RxDB, and I can't wait for others to be able to learn about and use it too.

When I was getting started using RxDB, there is SO MUCH great documentation, and you certainly have already written in detail about a lot of the questions that I had along the way. I personally felt like one of my biggest speed bumps was the structure of the topics in the left bar. I felt like they could be condensed and recategorized into more drop downs (like Migration and Server) so that information is easier to find. For instance, on the left hand side, after Migration, it feels like you've started to list a lot of features / topics that are important but they are kind of just added on and have no cohesion.

Also there's also a lot incredible information that you've put down in Articles that could be categorized with the relevant section so that people understand that there is more to read, and which topics those articles are related to. Just an idea - you could add a section at the end of RxDocuments called "Further Reading" that links to relevant articles, like attachments, migrations, etc.

As a somewhat new user who has just taken the hours to go through all the documentation you've written, I know that I'd be a great tester for clarity and organization! I would be genuinely more than happy to assist in any way that I can with this, whether that's coming up with visuals diagrams, proof reading individual pages for clarity, further help with organization, or anything else. 

Thank you!

**Comment by brunovinicius** [comment]
I found out that Angular PWA service worker tries to cache, intercept and respond to GraphQL requests and although throughout the tests it didn't seems to cause any issues, when we went to  prod we found out it was causing inconsistency on the data synced between clients (some documents missing on the local DB; distinct sets between them). Besides that we found out that it also duplicated pushes especially after recovering from offline or after refreshing an expired auth token. 

If you can see how this might cause issues, I would say its probably a good idea to add a warning box to the documentation calling attention to the fact. 

We added a header to our requests as documented on the link below, and it seemed to improve stability substantially.

https://angular.dev/ecosystem/service-workers/devops#bypassing-the-service-worker

**Comment by pubkey** [comment]
ping

**Comment by yashspr** [comment]
Thanks a lot of taking the initiative to improve the documentation. And, thanks a ton for building this amazing library. I have been building an offline first app, for which the replication module in this library is critical. Based on my usage so far, here are my suggestions:

1. More information on the flow of the replication. For example, when is the pull handler called, and when is the push handler called, what happens if there is a conflict from the pull handler, what happens when there is conflict from the push handler.

2. In continuation to point 1, what are the arguements to the conflict handler (assumed master state, real master state, new document state). I noticed during testing that the "assumed master state" which is passed to the conflict handler is sometimes present, but other times null. Why is that so?

To summarize, just providing more detailed information on replication protocol and the conflict handler and its flow would be of immese help.

**Comment by pubkey** [comment]
Ping

**Comment by gerardnll** [comment]
Hi! RxDB is such a great project and it's been so useful, I'm looking forward to go premium once I've done all the implementation. I've found some issues trying to do the following things and it's been difficult to find how to do them properly:

- How to 'truncate' collections but leave schema and everything intact. Is that possible? There are many functions but they are not clear enough for me. For example, I have N events and those events have N attendees. If I change active event, i want to clear the attendees collection so I can load the attendees for the next selected event. Should I create a DB for event? I don't want to leave traces everytime I change event.
- Is there a way to get information about replication/sync state? Changes not synced, syncing in progress, etc.
- Improve categorization of left bar topics.

**Comment by pubkey** [comment]
TODO https://discord.com/channels/969553741705539624/1027887523705389108/1311318952575500301

**Comment by tsekiguchi** [comment]
Hi pubkey et al,

I wanted to post here a first pass at the Table of Contents (sidebar) organization. Now that I'm pretty familiar with all the docs, I felt like placing information into some big buckets could help new users understand RxDB more quickly and more in depth, and help current users find what they're looking for. 

I've included all the items that you currently have in the sidebar, renamed a few for clarity, and suggested a couple new pages that might be useful.

Let me know what you think!

**Getting Started with RxDB**
- Introduction
- Quickstart
- Installation
- Development (renamed from Dev-Mode Plugin)
- Node.js Tutorial

**Why RxDB**
- Empowering Web Applications with Reactive Data Handling
- The Benefits of Browser Databases and RxDB
- Local First / Offline First (article)
- Downsides of Local First / Offline First
- Why NOSQL
- Alternatives

**Releases**

**How RxDB Works**
- Database Storage Architecture (renamed from RxStorage Layer)
- Transactions, Conflicts, & Revisions
- Replication Protocol
- What is a realtime database?
- JSON Database
- In Memory

**RxDB Architecture**
- RxStorage (free listed first, premium second)
    - Dexie.js
    - MongoDB
    - DenoKV
    - FoundationDB
    - LokiJS
    - Remote
    - Memory
    - IndexedDB
    - Memory
    - SQLite
    - Performance Comparisons (renamed from article, ‘**LocalStorage vs. IndexedDB vs. Cookies vs. OPFS vs. WASM-SQLite’)**
- Wrappers and Plugins
    - Workers
    - SharedWorkers
    - Encryption
    - Key Compression
    - Memory Mapped
    - Memory Synced
    - Sharding
    - Leader Election
    - Localstorage Meta Optimizer
    - CRDT
    - Electron
    - Create a plugin
- RxDatabase
- RxSchema & Collections
    - Schema Validation
    - Typescript Types
- RxDocument
    - Operations, including populate (which is currently its own page)
- Attachments
- Queries

**RxReplication**
- HTTP Replication
- RxDB Server Replication
- GraphQL Replication
- WebSocket Replication
- CouchDB Replication
- WebRTC Replication
- Firestore Replication
- NATS Replication

**RxDB Features**
- Migrations
    - Schema Migration
    - Adding Fields to your Schema (new page or section)
    - Storage Migration
- Pipelines
- Observables (new page)
- Schema Validation
- Cleanup
- Backup
- Encryption
- Key Compression
- Signals & Custom Reactivity
- RxState
- Local Documents
- Hooks (renamed from Middleware)
- Query Cache
- Full Text Search
- Query Optimizer
- Logger
- RxServer
    - RxServer
    - Scaling

**Examples**
- Embedded Database
- Flutter
- Mobile
- Javascript FE Database
- PWA
- React
- Capacitor
- Electron
- Ionic

**Performance Comparisons**
- Benchmarks
- Optimizations (new page that talks about strategies for getting the most out of RxDB)
- Why IndexedDB is slow
- WebSockets vs Server-Sent-Events vs Long-Polling vs WebRTC vs WebTransport

**FAQ**

**Contribute**

**Contact**
- Consulting
- Discord
- Email
- Chat

**Comment by EddiePengg** [comment]
Hi Rxdb The replication protocol of rxdb confuses me

**Comment by pubkey** [comment]
Closing this. I refactored most part and restructured the sidebar.
PRs to improve further are still welcome.
